### ISSUE TITLE
existing Icinga2 installation with PostgreSQL DB

### ISSUE THREAD

**Issue by SiliconAlley** [patrickpr/trapdirector#9]
**Main question**
We like to use trapdirector, but our installation is already done with Postgre-DB. Does trapdirector also work with it? Is there a possibility to translate the MySQL statements to Postres?
The ido-DB is in Postgre. As you create a schema, do we need MySQL in parallel for the trapdirector DB?
Does the configuration accept different databases? 

**Purpose**
We like to use trapdirector with our current Postgre-SQL installation.

**Your Environment:**
Postgresql 9.6
If related to the question in any way : 
latest Icinga Web 2 version and modules 2.7.1
latest Icinga 2 version (`icinga2 --version`): 2.10.5
Operating System and version: CentOS 7
Webserver, PHP versions: latest Apache, SCL PHP 7.1.8

Thanks
Thomas

**Comment by patrickpr** [comment]
Hi,
For now, I have only tested with MySQL but as I use PHP PDO, the plugin should be able to handle Postgre databases with minor updates.

There will also be a problem for now with the Postgre ido-DB.

Are you available to do some testing if I provide a patch ?

/Patrick

**Comment by SiliconAlley** [comment]
Hi Patrick,

thanks for the quick response. 
For sure we are available for some testing on our test machine. Let us know.

Thanks
Thomas

**Comment by patrickpr** [comment]
Hi,
I created a new branch (issue#9) for this, and let you know when available for testing.

**Comment by patrickpr** [comment]
Hi,

It was actually quite easy as I am using PDO, so I only had to do minor changes.

It works quite well on my lab, but I wasn't able to check the compatibility with postgresql IDO database.

Here is the test sequence : 

1) Download code from branch "Issue#9"

2) Follow the installation doc for this branch (https://github.com/patrickpr/trapdirector/blob/Issue%239/docs/01-install.md)

If you have a better and/or simpler way to create db & user for postGre than in the doc, please tell me !

3) With CLI, launch mib updates :  icingacli trapdirector mib update

Send me any errors shown

4) Test IDO database : 

On IcingaWeb2, go to Traps -> Handlers -> Add a trap handler

Enter a partial hostname : check that hostname(s) corresponding to the partial name are proposed.

Then, click Service and check the services for the host are displayed

5) Send a trap with the trapdirector/bin/test_trap_vX bash script and check they are in 'Trapd -> received'

6) Try to create a handler from this trap, save it, edit it, delete it....


If all this goes without error, next steps are for me to merge code and check non regression with mysql/mariadb .... and run postgreSQL version tests again :-)

/Patrick

**Comment by SiliconAlley** [comment]
Hi Patrick,

really nice. I will try this next week, I'm out for this week, but I'm really looking forward to test it.

Many thanks for your effort, I appreciate it.

Thomas

**Comment by SiliconAlley** [comment]
Hi Patrick,

unzipped/installed today, created new database, enabled permissions etc. so far. Then on the modules config page I managed to get the new DB with editing /etc/icingaweb2/resources.ini and saved. Short after this I also put in my API user (PW was in clear text in the webform) and I got kicked out.
The only thing I see now is:

Unable to authenticate, please check your API credentials

#0 /usr/share/icingaweb2/modules/trapdirector/library/Trapdirector/Icinga2Api.php(74): Icinga\Module\Trapdirector\Icinga2API->request(String, String, NULL, NULL)
#1 /usr/share/icingaweb2/modules/trapdirector/application/controllers/SettingsController.php(92): Icinga\Module\Trapdirector\Icinga2API->test(Array)
#2 /usr/share/icingaweb2/library/vendor/Zend/Controller/Action.php(507): Icinga\Module\Trapdirector\Controllers\SettingsController->indexAction()
#3 /usr/share/php/Icinga/Web/Controller/Dispatcher.php(76): Zend_Controller_Action->dispatch(String)
#4 /usr/share/icingaweb2/library/vendor/Zend/Controller/Front.php(937): Icinga\Web\Controller\Dispatcher->dispatch(Object(Icinga\Web\Request), Object(Icinga\Web\Response))
#5 /usr/share/php/Icinga/Application/Web.php(300): Zend_Controller_Front->dispatch(Object(Icinga\Web\Request), Object(Icinga\Web\Response))
#6 /usr/share/php/Icinga/Application/webrouter.php(99): Icinga\Application\Web->dispatch()
#7 /usr/share/icingaweb2/public/index.php(4): require_once(String)
#8 {main}

I don't get the config page back. Did a quick search, in the modules directory, but I don't know, where are the (wrong) credentials in now.

Continue tomorrow.
Thanks
Thomas

**Comment by patrickpr** [comment]
My bad for the error (I should read the function doc I wrote ). I updated the API tester. You can download it on Issue#9 branch : 
https://github.com/patrickpr/trapdirector/blob/Issue%239/library/Trapdirector/Icinga2Api.php 
and replace your current file located at 
/usr/share/icingaweb2/modules/trapdirector/library/Trapdirector/

In case anythink like this happens again, you can erase (or edit) the /etc/icingaweb2/modules/trapdirector/config.ini file.

Note : you can normally create a db resource in web GUI at /icingaweb2/config/resource, is there a specific reason you edited the resource.ini file instead ?

Patrick

**Comment by SiliconAlley** [comment]
Hi Patrick,

updated the Icinga2Api.php file and deleted the config.ini file. Thanks, no problem.

For creating the db resource I couldn't select my new db "icinga_itd" in the web GUI and I knew about the resource.ini file, put it there and I could choose it, so it was fine for me.

Now I try to get the schema implemented into the new db, but it doesn't work with the link "Create Schema" nor importing on the command line.
# psql -U icinga_itd -d icinga_itd < /usr/share/icingaweb2/modules/trapdirector/SQL/schema_v1.pgsql

I noticed in the pgsql file #PREFIX# which should be "traps_" from the config page. So I get complaints on the console like:
ERROR:  syntax error at or near "#"
LINE 1: CREATE TABLE #PREFIX#db_config (
Sure, the schema file doesn't know, what prefix is configured on the config page.

Following the Create Schema link brings me to this Error:

Uncaught Error: Class 'Icinga\Module\Trapdirector\Icinga2API' not found in /usr/share/icingaweb2/modules/trapdirector/bin/trap_class.php:234
Stack trace:
#0 /usr/share/icingaweb2/modules/trapdirector/bin/trap_class.php(137): Trap->getAPI()
#1 /usr/share/icingaweb2/modules/trapdirector/bin/trap_class.php(64): Trap->getOptions()
#2 /usr/share/icingaweb2/modules/trapdirector/application/controllers/SettingsController.php(193): Trap->__construct('/etc/icingaweb2')
#3 /usr/share/icingaweb2/library/vendor/Zend/Controller/Action.php(507): Icinga\Module\Trapdirector\Controllers\SettingsController->createschemaAction()
#4 /usr/share/php/Icinga/Web/Controller/Dispatcher.php(76): Zend_Controller_Action->dispatch('createschemaAct...')
#5 /usr/share/icingaweb2/library/vendor/Zend/Controller/Front.php(937): Icinga\Web\Controller\Dispatcher->dispatch(Object(Icinga\Web\Request), Object(Icinga\Web\Response))
#6 /usr/share/php/Icinga/Application/Web.php(300): Zend_Controller_Front->dispatch(Object(Icinga\Web\Request), Object(I

#0 [internal function]: Icinga\Application\Web->Icinga\Application\{closure}()
#1 {main}

Is it user related, when icingaweb accesses the database?

To complete, my Configuration page now shows:

<green>Database set in config.ini

<green>Database exists in Icingaweb2 config

<yellow>Database does not exist/invalid credentials/no schema : icinga_itd, Message : SQLSTATE[42P01]: Undefined table: 7 ERROR: relation "traps_db_config" does not exist LINE 1: SELECT c.value FROM traps_db_config AS c WHERE (name='db_ver... ^, query was: SELECT c.value FROM traps_db_config AS c WHERE (name='db_version')

Configure it here : /trapdirector/settings/createschema

<green>IDO Database OK

<yellow>API parameters not configured

<yellow>Port UDP/162 is not open : snmptrapd must not be started

Thanks for advise
Thomas

**Comment by patrickpr** [comment]
Hi,
Problem is the same than Issue#8 : incorrect values in ini file generates runtime errors. Can you setup API user (don't forget to reload icinga2) ? Or try removing in config.ini.

Notes : as I allow user to set table prefix, you need to use the GUI to create the schema : this is why I have the '#PREFIX#' everywhere in SQL files.

**Comment by SiliconAlley** [comment]
Hi Patrick,

I created already an API user, when I put this user in, I get:

Uncaught Error: Class 'Icinga\Module\Trapdirector\Icinga2Api' not found in /usr/share/icingaweb2/modules/trapdirector/application/controllers/SettingsController.php:89

With removing in config.ini all the icingaAPI config parameters I still get the error:

Uncaught Error: Class 'Icinga\Module\Trapdirector\Icinga2API' not found in /usr/share/icingaweb2/modules/trapdirector/bin/trap_class.php:234

Permissions where right:
-rw-r--r--. 1 root root 188711 Sep 17 07:01 Icinga2Api.php

Do the above errors complain about the Icinga2Api.php?

I also removed the entry in resource.ini and created it over the webGUI again.

The problem in Issue#8 is not API related, is it the same? You must know. ;-)

I still think it might be user related, as asked above. User "apache" has to put the schema in the db, maybe I think the wrong way.

Thanks
Thomas

**Comment by patrickpr** [comment]
Hi,

I noticed the size of Icinga2Api.php is about 188  kB : it should be around 12 kB. Could you check if the file is correct ?

- The problem is php files (settings and trapclass) don't find the Icinga2API class which should be defined in Icinga2Api.php

- It can be related to issue #8 as there is the same type of error : class not found - although it is defined in files - when there is errors in config.ini file. (But I'm not 100% sure).

- It is not user apache which put the schema in the database : the module gets the user defined in resource.ini to connect and push the schema : you can validate user can connect when you edit the ressource of icingaweb2 (/icingaweb2/config/resource ) : there is a "validate configuration" button when editing a resource. 

- If the Icinga2Api.php file is correct, I will try to reproduce the error (I've not been able to do so for now).

Regards,
Patrick

**Comment by SiliconAlley** [comment]
Hi,

sorry for the trouble, it was wget. There was a lot of html in the php file in. I was also wondering about the file size. Now with the raw file it looks like this:
-rw-r--r--. 1 root root 12385 Sep 18 06:57 Icinga2Api.php

And the schema creation is successfull
 SCHEMA CREATION
Creating schema :

[2019/09/18 09:02:11] [trap_class.php] : Creating : table traps_db_config
[2019/09/18 09:02:11] [trap_class.php] : Creating : secret SQL stuff :-)
[2019/09/18 09:02:11] [trap_class.php] : Creating : secret SQL stuff :-)
[2019/09/18 09:02:11] [trap_class.php] : Creating : table traps_mib_cache
[2019/09/18 09:02:11] [trap_class.php] : Creating : table traps_mib_cache_syntax
[2019/09/18 09:02:11] [trap_class.php] : Creating : table traps_mib_cache_tc
[2019/09/18 09:02:12] [trap_class.php] : Creating : table traps_mib_cache_trap_object
[2019/09/18 09:02:12] [trap_class.php] : Creating : secret SQL stuff :-)
[2019/09/18 09:02:12] [trap_class.php] : Creating : table traps_received
[2019/09/18 09:02:12] [trap_class.php] : Creating : table traps_received_data
[2019/09/18 09:02:12] [trap_class.php] : Creating : table traps_rules
[2019/09/18 09:02:12] [trap_class.php] : Schema created

Now I will continue with the API setup...

Many thanks
Good work
Thomas

**Comment by patrickpr** [comment]
I'm merging Issue#9 branch into master as mysql seems to be working OK, so future changes will be on master and Issue#9 branch will be deleted.

**Comment by SiliconAlley** [comment]
Hi Patrick,
OK, ^^ Postgresql you mean...

When setting the API I had to put the API user also to /etc/icinga2/features-available/api.conf
I had this issue before too, with the "normal" API user. This seems not to be on every system.
I checked all the time with icinga2 daemon -C and got all the time "Instantiated 1 ApiUser", which was the already existing user. Just FYI, not Trapdirector related.

One related thing is, that on the configuration page the API password is not masked. In the "icinga2 API password" field is "cleartextpassword"


When now using the plugin I get some SQL errors.
When clicking "Traps" - "Hosts" or "Received" - "Hosts" I get: 2 times

SQLSTATE[42803]: Grouping error: 7 ERROR: column "t.source_name" must appear in the GROUP BY clause or be used in an aggregate function
LINE 1: SELECT t.source_name, t.source_ip, t.trap_oid, count(*) AS c...
^, query was: SELECT t.source_name, t.source_ip, t.trap_oid, count(*) AS count, UNIX_TIMESTAMP(max(t.date_received)) AS last_sent FROM traps_received AS t GROUP BY t.source_ip,
t.trap_oid ORDER BY t.source_ip ASC LIMIT 25

#0 /usr/share/icingaweb2/library/vendor/Zend/Db/Statement.php(297): Zend_Db_Statement_Pdo->_execute(Array)
#1 /usr/share/icingaweb2/library/vendor/Zend/Db/Adapter/Abstract.php(470): Zend_Db_Statement->execute(Array)
#2 /usr/share/icingaweb2/library/vendor/Zend/Db/Adapter/Pdo/Abstract.php(238): Zend_Db_Adapter_Abstract->query(String, Array)
#3 /usr/share/icingaweb2/library/vendor/Zend/Db/Adapter/Abstract.php(725): Zend_Db_Adapter_Pdo_Abstract->query(Object(Zend_Db_Select), Array)
#4 /usr/share/icingaweb2/modules/trapdirector/library/Trapdirector/Tables/TrapTableHostList.php(175): Zend_Db_Adapter_Abstract->fetchAll(Object(Zend_Db_Select))
#5 /usr/share/icingaweb2/modules/trapdirector/library/Trapdirector/Tables/TrapTableHostList.php(80): Icinga\Module\TrapDirector\Tables\TrapTableHostList->getTable()
#6 zend.view:///usr/share/icingaweb2/modules/trapdirector/application/views/scripts/received/hosts.phtml(14): Icinga\Module\TrapDirector\Tables\TrapTableHostList->render()
#7 /usr/share/php/Icinga/Web/View.php(262): include(String)
#8 /usr/share/icingaweb2/library/vendor/Zend/View/Abstract.php(877): Icinga\Web\View->_run(String)
#9 /usr/share/icingaweb2/library/vendor/Zend/Controller/Action/Helper/ViewRenderer.php(904): Zend_View_Abstract->render(NULL)
#10 /usr/share/icingaweb2/library/vendor/Zend/Controller/Action/Helper/ViewRenderer.php(925): Zend_Controller_Action_Helper_ViewRenderer->renderScript(String, NULL)
#11 /usr/share/icingaweb2/library/vendor/Zend/Controller/Action/Helper/ViewRenderer.php(964): Zend_Controller_Action_Helper_ViewRenderer->render()
#12 /usr/share/icingaweb2/library/vendor/Zend/Controller/Action/HelperBroker.php(272): Zend_Controller_Action_Helper_ViewRenderer->postDispatch()
#13 /usr/share/icingaweb2/library/vendor/Zend/Controller/Action.php(518): Zend_Controller_Action_HelperBroker->notifyPostDispatch()
#14 /usr/share/php/Icinga/Web/Controller/Dispatcher.php(76): Zend_Controller_Action->dispatch(String)
#15 /usr/share/icingaweb2/library/vendor/Zend/Controller/Front.php(937): Icinga\Web\Controller\Dispatcher->dispatch(Object(Icinga\Web\Request), Object(Icinga\Web\Response))
#16 /usr/share/php/Icinga/Application/Web.php(300): Zend_Controller_Front->dispatch(Object(Icinga\Web\Request), Object(Icinga\Web\Response))
#17 /usr/share/php/Icinga/Application/webrouter.php(99): Icinga\Application\Web->dispatch()
#18 /usr/share/icingaweb2/public/index.php(4): require_once(String)
#19 {main}

If you like, you are invited for an online session to see it directly on the system.

Thanks
Thomas

**Comment by patrickpr** [comment]
(No, I mean MySQL : I did some test to see if what I have done for PostGre didn't break the MySQL functions)

When doing some further tests, I also noticed filters and grouping (by host in your case) didn't work : I need to use specific PostgreSQL syntax -> I'll work on this and let you know.

Patrick

**Comment by patrickpr** [comment]
Hi,
It is a problem with SQL queries when grouping by host (or when using filters) : I'm working on this this week.

Are all the other elements working OK ?

thanks for testing !

Patrick

**Comment by SiliconAlley** [comment]
Hi Patrick,

many thanks for your efforts.
Yes, we are testing and get the traps, we have to specify the handlers to get some events in Icinga etc.

I'm just dealing with SELinux on CentOS7. The webinterface cannot detect the snmptrapd process in enforcing mode, and shows a yellow warning. In permissive mode all shows green on the configuration page.

Unfortunately audit2allow doesn't detect anything, so it's hard for me to go behind. "httpd_enable_cgi --> on" was already set.
"aureport -a" shows something. I will investigate... It's a bit offtopic here.

Also, when changing "Log destination" under "Status & Mibs" to "file", it does not change and fires a red lower bar with "error[object Object]" even in permissive mode.

Thanks again
Thomas

**Comment by patrickpr** [comment]
Hi,

I suspected there would be problems with the snmptrapd testing on systems with security on : I will try to test the mode to return a specific warning with SELinux enforcing mode.
There will be a script (ongoing dev on 1.1 branch) so you can check all of this with a bash script and be able to test with SELinux on.

- I updated master branch so trap list works with postgre : 'hosts' tab in 'received' and filters on trap list are now working with postgre
- The error in log destination (among others)  was reported in issue#10 and corrected today (in master branch too).

/Patrick

**Comment by SiliconAlley** [comment]
Hi from home,

thanks, I will try tomorrow and looking forward to.
Just to let you know, snmptrapd works, we get traps in under received etc., even in enforcing mode. Only the GUI shows yellow and cannot get the process from ps
I wanted to investigate the php-files tomorrow and see, how you check for the process. First lets update.
I found many complaints searching the web, that the audit.log is empty, when php checks the process and cannot in enforcing mode.

Have it now on 2 machines, they behave equal.

Thomas

**Comment by SiliconAlley** [comment]
Hi Patrick,

updated the modules directory, all files/folders have date stamp Sep 24 20:53 and "Log destination" can be changed now.

But, I get the exact same error message (SQLSTATE[42803]: Grouping error: 7 ERROR:...) when going to "Hosts" like in https://github.com/patrickpr/trapdirector/issues/9#issuecomment-533008005

While writing this (and in permissive mode) I'm thinking about permissions for the folders in the error message, especially /usr/share/icingaweb2/library/vendor/Zend/..
They are all set to root:root with 644, hmm, read for others should be OK, I guess.

Another thing I see in the SQL-statement, but you know better:
query was: SELECT t.source_name, t.source_ip, t.**trap_**oid,

I left the table prefix "traps_", so I'm wondering, if all tables should have trap_ in table name.
Just my thoughts, skip, if I'm wrong.

Thanks
Thomas

**Comment by patrickpr** [comment]
OK, I'll check my updates (I don't have much time these days so probably not before friday).
/Patrick

**Comment by patrickpr** [comment]
I checked the commits and all seems OK, as well as my lab with postgre showing host correctly.

- Can you check if the TrapTableHostList.php file line 184 is : 
)->group(array('t.source_ip','t.source_name','t.trap_oid'))

- Check if the SQL query shown when on "host" has the GROUP BY  with all three columns : 
GROUP BY  t.source_ip, t.source_name, t.trap_oid

- Run the query in cli with psql and report back the error.

Thanks for testing !
Patrick

**Comment by SiliconAlley** [comment]
Hi Patrick,

first, it works now, many thanks!

I downloaded the master.zip again and noticed the files in are newer:
-rw-r--r--. 1 root root 6650 Sep 27 12:59 TrapTableHostList.php

Now reading the error above it is clearly in front of my eyes. "ERROR: column "t.source_name" must appear in the GROUP BY clause.."

The GROUP BY was missing the middle part "t.source_name". It was still )->group(array('t.source_ip','t.trap_oid'), even in the  Sep 24 20:53 release.
-rw-r--r--. 1 root root 6636 Sep 24 20:53 TrapTableHostList.php

I'll continue next week further testing, but I think I can close this issue.

Again, many thanks for your quick and friendly help.

Thomas

**Comment by patrickpr** [comment]
Hi,
Glad it works !
Did the version pass all test I listed in https://github.com/patrickpr/trapdirector/issues/9#issuecomment-529944166 ?

I will close this case next week if no other problems on pgsql arise.

Thanks to you for testing.

Patrick

**Comment by SiliconAlley** [comment]
Hi Patrick,

yes, all is working now, I would say. From Configuration page: All tests OK : DB, IdoDB, API, snmptrapd

Still some SELinux issues, but this is not module related I guess.

Uploaded existing mibs and "update mib db" took 7-8 hours.

If you like I can share my install documentation.

Many thanks.
Thomas

**Comment by patrickpr** [comment]
Hi,

Nice to see all is working.
Mib updates takes way too much time, but I'm working on it.
I would very much like your install doc so I can adjust mine : if you want to keep it non public, send it to me by email.

Regards,
Patrick

**Comment by SiliconAlley** [comment]
Hi,

sure, I will create a pdf out of the pages, then send it by email (if I find an address).

Thanks
Thomas

**Comment by patrickpr** [comment]
Great, use this one : patrick at proy.org

**Comment by SiliconAlley** [comment]
Hi,

you can remove your address here ^^ for obscurity, have it in my inbox.
I will send it next week, tomorrow is public holiday here in Germany.

Thanks
Thomas

**Comment by patrickpr** [comment]
I will close the issue, feel free to reopen if you have some issues related to his or another one.

/Patrick

**Comment by IfrahKomal** [comment]
Hi,
I am unable to solve this error .  

SQLSTATE[42S02]: Base table or view not found: 1146 Table 'traptest.traps_db_config' doesn't exist

#0 /usr/share/icingaweb2/modules/trapdirector/library/Trapdirector/TrapsProcess/TrapConfig.php(97): PDO->query()
#1 /usr/share/icingaweb2/modules/trapdirector/library/Trapdirector/TrapsProcess/TrapConfig.php(84): Trapdirector\Trap->getDBConfig()
#2 /usr/share/icingaweb2/modules/trapdirector/library/Trapdirector/TrapsProcess/TrapConfig.php(65): Trapdirector\Trap->getDBConfigIfSet()
#3 /usr/share/icingaweb2/modules/trapdirector/library/Trapdirector/TrapsProcess/Trap.php(129): Trapdirector\Trap->getDatabaseOptions()
#4 /usr/share/icingaweb2/modules/trapdirector/application/controllers/SettingsController.php(369): Trapdirector\Trap->__construct()
#5 /usr/share/icingaweb2/library/vendor/Zend/Controller/Action.php(507): Icinga\Module\Trapdirector\Controllers\SettingsController->createschemaAction()
#6 /usr/share/php/Icinga/Web/Controller/Dispatcher.php(76): Zend_Controller_Action->dispatch()
#7 /usr/share/icingaweb2/library/vendor/Zend/Controller/Front.php(937): Icinga\Web\Controller\Dispatcher->dispatch()
#8 /usr/share/php/Icinga/Application/Web.php(304): Zend_Controller_Front->dispatch()
#9 /usr/share/php/Icinga/Application/webrouter.php(107): Icinga\Application\Web->dispatch()
#10 /usr/share/icingaweb2/public/index.php(4): require_once(String)
#11 {main}

**Comment by Fl33Tr** [comment]
> Hi, I am unable to solve this error .
> 
> SQLSTATE[42S02]: Base table or view not found: 1146 Table 'traptest.traps_db_config' doesn't exist
> 
> #0 /usr/share/icingaweb2/modules/trapdirector/library/Trapdirector/TrapsProcess/TrapConfig.php(97): PDO->query() #1 /usr/share/icingaweb2/modules/trapdirector/library/Trapdirector/TrapsProcess/TrapConfig.php(84): Trapdirector\Trap->getDBConfig() #2 /usr/share/icingaweb2/modules/trapdirector/library/Trapdirector/TrapsProcess/TrapConfig.php(65): Trapdirector\Trap->getDBConfigIfSet() #3 /usr/share/icingaweb2/modules/trapdirector/library/Trapdirector/TrapsProcess/Trap.php(129): Trapdirector\Trap->getDatabaseOptions() #4 /usr/share/icingaweb2/modules/trapdirector/application/controllers/SettingsController.php(369): Trapdirector\Trap->__construct() #5 /usr/share/icingaweb2/library/vendor/Zend/Controller/Action.php(507): Icinga\Module\Trapdirector\Controllers\SettingsController->createschemaAction() #6 /usr/share/php/Icinga/Web/Controller/Dispatcher.php(76): Zend_Controller_Action->dispatch() #7 /usr/share/icingaweb2/library/vendor/Zend/Controller/Front.php(937): Icinga\Web\Controller\Dispatcher->dispatch() #8 /usr/share/php/Icinga/Application/Web.php(304): Zend_Controller_Front->dispatch() #9 /usr/share/php/Icinga/Application/webrouter.php(107): Icinga\Application\Web->dispatch() #10 /usr/share/icingaweb2/public/index.php(4): require_once(String) #11 {main}

Hello IfrahKomal,
I am new to trapdirector myself. I think this error states a missing database table.
My best guess is that you are either

1) Missing a database for trapdirector to write into 
or
2) Selecting the wrong database in the trapdirector settings
